core/chains/solana/soltxm/txm.go

```diff
@@ -25,82 +34,368 @@ var (
+			}
+
+			// send tx + signature to simulation queue
+			msg.signature = sig
+			select {
+			case txm.chSim <- msg:
+			default:
+				txm.lggr.Warnw("failed to enqeue tx for simulation", "queueFull", len(txm.chSend) == MaxQueueLen, "tx", msg)
+			}
+
+			txm.lggr.Debugw("transaction sent", "signature", sig.String())
+		case <-txm.chStop:
+			return
+		}
+	}
+}
+
+func (txm *Txm) sendWithRetry(chanCtx context.Context, tx *solanaGo.Transaction, timeout time.Duration) (solanaGo.Signature, error) {
+	// fetch client
+	client, err := txm.client.Get()
+	if err != nil {
+		return solanaGo.Signature{}, errors.Wrap(err, "failed to get client in soltxm.sendWithRetry")
+	}
+
+	// create timeout context
+	ctx, cancel := context.WithTimeout(chanCtx, timeout)
+
+	// send initial tx (do not retry and exit early if fails)
+	sig, err := client.SendTx(ctx, tx)
+	if err != nil {
+		cancel()            // cancel context when exiting early
+		txm.txs.Failed(sig) // increment failed metric
+		return solanaGo.Signature{}, errors.Wrap(err, "tx failed initial transmit")
+	}
+
+	// store tx signature + cancel function
+	if err := txm.txs.Insert(sig, cancel); err != nil {
+		cancel() // cancel context when exiting early
+		return solanaGo.Signature{}, errors.Wrapf(err, "failed to save tx signature (%s) to inflight txs", sig)
+	}
+
+	// retry with exponential backoff
+	// until context cancelled by timeout or called externally
+	go func() {
+		deltaT := 1 // ms
+		tick := time.After(0)
+		for {
+			select {
+			case <-ctx.Done():
+				// stop sending tx after retry tx ctx times out (does not stop confirmation polling for tx)
+				txm.lggr.Debugw("stopped tx retry", "signature", sig)
+				return
+			case <-tick:
+				go func() {
+					retrySig, err := client.SendTx(ctx, tx)
+					// this could occur if endpoint goes down or if ctx cancelled
+					if err != nil {
+						if strings.Contains(err.Error(), "context canceled") || strings.Contains(err.Error(), "context deadline exceeded") {
+							txm.lggr.Debugw("ctx error on send retry transaction", "error", err, "signature", sig)
+						} else {
+							txm.lggr.Warnw("failed to send retry transaction", "error", err, "signature", sig)
+						}
+						return
+					}
+					// this should never happen
+					if retrySig != sig {
+						txm.lggr.Criticalw("original signature does not match retry signature", "expectedSignature", sig, "receivedSignature", retrySig)
+					}
+				}()
+			}
+
+			// exponential increase in wait time, capped at 500ms
+			deltaT *= 2
+			if deltaT > MaxRetryTimeMs {
+				deltaT = MaxRetryTimeMs
+			}
+			tick = time.After(time.Duration(deltaT) * time.Millisecond)
+		}
+	}()
+
+	// return signature for use in simulation
+	return sig, nil
+}
+
+// goroutine that polls to confirm implementation
+// cancels the exponential retry once confirmed
+func (txm *Txm) confirm(ctx context.Context) {
+	defer txm.done.Done()
+
+	timestamps := map[solanaGo.Signature]time.Time{}
```

Comment:
We have a component that holds our `PendingTxContext`, and as timestamps are part of that context that we want to keep it might be cleaner to add time tracking there.

A change like that would make this code much cleaner, less prone to bugs.